Changes to train_data_tokenize.py and est_data_tokenize.py

shashwat623 opened the pull request:
**train_data_tokenize.py**

The training data from kaggle dataset contains 2 rows in which the "question2" column has NAN values.
 Without replacing NAN values with empty "" , the code gives an error while executing this file as various functions like word.strip() cannot work on NAN values. Performing the commited change solves the issue.

**test_data_tokenize.py**

Since _self.map_word_index = pickle.load(open('preproc/map_word_index.pkl', 'r'))_  is being called multiple times for each cpu thread allocation, the above pickle model is only unpickling the first time it is being called, for rest calls it is showing EOF(End of file error).
This is because Pickle keeps track of the objects it has already serialized, so later references to the same object won’t be serialized again.
This issue was solved by unplickling _map_word_index.pkl_ initially in a global scope and assigning this unpickled object to _self.map_word_index_  for every thread it is being called for.
Hope this helps. Please do let me know.
![Screenshot (40)](https://user-images.githubusercontent.com/42846105/79852955-0867ef00-83e5-11ea-8f04-6f8c1eb1e48d.png)

shashwat623 committed:
Update train_data_tokenize.py

The training data from kaggle dataset contains 2 rows in which the "question2" column has NAN values.
 Without replacing NAN values with empty "" , the code gives an error while executing this file as various functions like word.strip() cannot work on NAN values. Performing the commited change solves the issue.

shashwat623 committed:
Update test_data_tokenize.py

Since **self.map_word_index = pickle.load(open('preproc/map_word_index.pkl', 'r'))**  is being called multiple times for each cpu thread allocation, the above pickle model is only unpickling the first time it is being called, for rest calls it is showing EOF(End of file error).
This is because Pickle keeps track of the objects it has already serialized, so later references to the same object won’t be serialized again.
This issue was solved by unplickling **map_word_index.pkl** initially in a global scope and assigning this unpickled object to 
**self.map_word_index**  for every thread it is being called for.
Hope this helps.

ameyagodbole commented:
@shashwat623 thanks for your interest in our project. I have a couple of questions though:
1. [this line](https://github.com/dalmia/Quora-Question-Pairs/blob/1b747a9c7c63e7e81b0763b46c0c4847ca350449/preproc/test_data_tokenize.py#L22) looks problematic since you are calling load on an object. Have you tested this by running it?
2. Do you think this issue might be specific to IPython? AFAI remember we did not run into it. Maybe @dalmia can correct me

shashwat623 commented:
@ameyagodbole Thank you for your reply.

1. I know it would have been more appropriate to use pickle.loads(obj) instead of pickle.load(obj) as the latter is meant to load pickled data from a file-like object, but for newer pickle versions either one of them works equally good. Maybe they have overloaded the functions in the newer versions. Anyways, to cut down ambiguity I'll re commit load() to loads().
Yes, I did test it by running and it runs perfectly good.

2. I am not sure whether the issue is specific to IPython or not, in my case it gave the same error on both Google colab and local runtime. My friend faced the same issue on spyder.

3. Since self.map_word_index = pickle.load(open('preproc/map_word_index.pkl', 'r')) is being called multiple times for each chunk of data allocation, the above pickle model is only unpickling the first time it is being called, for rest calls it is showing EOF(End of file error). 
I believe it is because because Pickle keeps track of the objects it has already serialized, so later references to the same object won’t be serialized again. In other words for a single runtime, by default, objects will only be serialized once. The model should be unpickled once globally, this would also be more computationally efficient. Refer to my output image in the previous comment, it only unpickles for the first call and shows EOF error thenceforth.
Take a look at the Pickle documentation for more info. You will notice while unpickling, the load()/loads() function takes reference from how the data was originally dumped, that is, if the data was dumped in successive batches or all at once. It will use the same mapping to load the data while unpickling. In the code, the data is dumped  all at once so it will load the data all at once and only once.
 I have attached some screenshots of the documents and codes for better reference.
Hope it helps. Thank you!
![Screenshot (43)](https://user-images.githubusercontent.com/42846105/80285394-41d08f80-8742-11ea-9e86-9b1496d9af90.png)
![Screenshot (44)](https://user-images.githubusercontent.com/42846105/80285396-45641680-8742-11ea-8977-04c186d317eb.png)

shashwat623 committed:
Update test_data_tokenize.py

Changed pickle.load() to pickle.loads() as loads() can be called on an object. In the newer Pickle versions so can load()

ameyagodbole commented:
@shashwat623 sorry for the delay in replying. My original issue with `load` was that you were calling pickle.load(obj) instead of pickle.load(file_obj). Why does this work? According to the documentation, it should be a file_obj that has `read()` attribute

shashwat623 committed:
Update test_data_tokenize.py

shashwat623 commented:
@ameyagodbole sorry, this was my mistake. I did not refer to my executed code while commiting in the repo so made a blunder.  Yes, I **cannot** call `pickle.load(map_word_index_model)`.
I had actually called `model=pickle.load(file_obj)` in the executed code and then simply assigned the loaded model as `self.map_word_index = model`
All I'm trying to say is that we need to load the pickle model only once and not for every batch.
I have made the changes in the latest commit. Please do have a look and let me know.

ameyagodbole commented:
Squashed and merged. Thanks

shashwat623 commented:
Glad I could contribute. Thanks!!
